src/fromager/resolver.py

```diff
@@ -180,11 +181,42 @@ def resolve_from_provider(
     raise ValueError(f"Unable to resolve {req}")
 
 
+class RetrieveMethod(enum.StrEnum):
+    tarball = "tarball"
+    git_https = "git+https"
+    git_ssh = "git+ssh"
+
+    @classmethod
+    def from_url(cls, download_url) -> tuple[RetrieveMethod, str, str | None]:
+        """Parse a download URL into method, url, reference"""
+        scheme, netloc, path, query, fragment = urlsplit(
+            download_url, allow_fragments=False
+        )
+        match scheme:
+            case "https":
+                return RetrieveMethod.tarball, download_url, None
+            case "git+https":
+                method = RetrieveMethod.git_https
+            case "git+ssh":
+                method = RetrieveMethod.git_ssh
+            case _:
+                raise ValueError(f"unsupported download URL {download_url!r}")
+        # remove git+
+        scheme = scheme[4:]
+        # split off @ revision
+        if "@" not in path:
+            raise ValueError(f"git download url {download_url!r} is missing '@ref'")
+        path, ref = path.rsplit("@", 1)
+        return method, urlunsplit((scheme, netloc, path, query, fragment)), ref
+
+
 def get_project_from_pypi(
     project: str,
     extras: typing.Iterable[str],
     sdist_server_url: str,
     ignore_platform: bool = False,
+    *,
+    override_download_url: str | None = None,
```

Comment:
It's not clear how a caller would know what an alternative download site might be, and why that would be anything other than another package index server that could have been passed as the sdist_server_url parameter. What's the use case for using on package index to resolve a version and another server to act as the source of the download? Shouldn't the resolver look at the place where it's going to download the package to see what versions are available there?